### PR TITLE
Implement resolvePermissions

### PR DESCRIPTION
**Pull request by mateuszradomski**
Resolves L2B-6047

### CODE REVIEW

File: packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts [hunk 1]
```diff
@@ -0,0 +1,818 @@
...
+    const vault = EthereumAddress.random()
+
+    const graph: Graph = {
+      nodes: [
+        { address: actorA, members: [], threshold: 1 },
+        { address: actorB, members: [], threshold: 1 },
+        { address: actorC, members: [], threshold: 1 },
+        { address: msig, members: [0, 1, 2], threshold: 2 },
+        { address: vault, members: [], threshold: 1 },
+      ],
```
Comment (sz-piotr) [inline packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts:309]:
```suggestion
        { address: msig, multisig: {
          members: [0, 1, 2],
          threshold: 2,
          delay: 123,
        } },
```

File: packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts [hunk 2]
```diff
@@ -0,0 +1,818 @@
...
+          { address: vault, delay: 0 },
+          { address: proxy, delay: 0 },
+          { address: timelockB, delay: 0 },
+          { address: actor, delay: 100 },
+        ],
+      },
+    ])
+  })
+
+  it('three actors, four contracts, three multisig with members, mixed threshold', () => {
```
Comment (sz-piotr) [inline packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts:243]:
```suggestion
  it('three actors, four contracts, three multisigs with members, mixed threshold', () => {
```

File: packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+        type: 'configure',
+        path: [
+          { address: zkSync, delay: 0 },
+          { address: securityCouncil, delay: 0 },
+        ],
+      },
+    ])
+  })
+
+  it('one actor, four contracts, two timelocks with same delays', () => {
```
Comment (sz-piotr) [inline packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts:161]: Edge cases:
Two upgrades same delay, different path
Two upgrades different delay, different path
Configure longer delay than upgrade (collapsed?)
Configure same delay different path (collapsed or not?)

File: packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+
+    const graph: Graph = {
+      nodes: [
+        { address: verifier, members: [], threshold: 1 },
+        { address: governance, members: [], threshold: 1 },
+        { address: zkSync, members: [], threshold: 1 },
+        { address: upgradeGatekeeper, members: [], threshold: 1 },
+        { address: zkSyncMsig, members: [], threshold: 1 },
+        { address: securityCouncil, members: [], threshold: 1 },
+      ],
```
Comment (sz-piotr) [inline packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts:116]:
```suggestion
        { address: verifier, members: [], threshold: 1 },
        { address: governance, members: [], threshold: 1 },
        { address: zkSync, members: [5, 6], threshold: 2 },
        { address: upgradeGatekeeper, members: [], threshold: 1 },
        { address: zkSyncMsig, members: [], threshold: 1 },
        { address: actorA, members: [], threshold: 1 },
        { address: actorB, members: [], threshold: 1 },
```

File: packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+          { address: zkSyncMsig, delay: 21 },
+        ],
+      },
+      {
+        type: 'configure',
+        path: [
+          { address: zkSync, delay: 0 },
+          { address: securityCouncil, delay: 0 },
+        ],
+      },
```
Comment (sz-piotr) [inline packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts:157]:
```suggestion
      {
        type: 'configure',
        path: [
          { address: zkSync, delay: 0 }, // contract
          { address: zkSync, delay: 0 }, // embedded security council
        ],
      },
```

**Review by sz-piotr — CHANGES_REQUESTED** [review]

File: packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts [hunk 6]
```diff
@@ -0,0 +1,1096 @@
...
+      {
+        type: 'configure',
+        path: [
+          { address: zkSync, delay: 0 }, // contract
+          { address: zkSync, delay: 0 }, // embedded security council
+        ],
+      },
+    ])
+  })
+
```
Comment (sz-piotr) [inline packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts:164]: Tests to add:

A->A (zksync lite simplified)
A->A->B
A->B->A
A->B->A->C->A (ignore loop - a upgrade b, b act a, a act c, c act a)

File: packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts [hunk 7]
```diff
@@ -0,0 +1,1096 @@
...
+  it('one actor, four contracts, two timelocks with same delays', () => {
+    const actor = EthereumAddress.random()
+    const timelockA = EthereumAddress.random()
+    const timelockB = EthereumAddress.random()
+    const proxy = EthereumAddress.random()
+    const vault = EthereumAddress.random()
+
+    const graph: Graph = {
+      nodes: [
+        { address: actor, threshold: 1 },
```
Comment (sz-piotr) [inline packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts:174]:
```
node(actor),
node(timelock),
node(multisig, { threshold: 2 }),
node(multisig2, { threshold: 3, delay: 1234 })
```

File: packages/discovery/src/discovery/permission-resolving/resolvePermissions.ts [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+export interface GrantedPermission {
+  type: Permission
+  path: PathElement[]
+}
+
+export function resolvePermissions(graph: Graph): GrantedPermission[] {
+  const result: GrantedPermission[] = []
+
+  for (let nodeId = 0; nodeId < graph.nodes.length; nodeId += 1) {
```
Comment (sz-piotr) [inline packages/discovery/src/discovery/permission-resolving/resolvePermissions.ts:60]:
```suggestion
  for (const [i, node] of graph.nodes.entries()) {
```

File: packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts [hunk 9]
```diff
@@ -0,0 +1,1096 @@
...
+      edges: [
+        { type: 'upgrade', fromNode: 0, toNode: 3, delay: 0 },
+        { type: 'upgrade', fromNode: 1, toNode: 3, delay: 0 },
+        { type: 'upgrade', fromNode: 2, toNode: 3, delay: 0 },
+        { type: 'act', fromNode: 3, toNode: 4, delay: 21 },
+        { type: 'configure', fromNode: 2, toNode: 2, delay: 0 },
+        { type: 'member', fromNode: 2, toNode: 5, delay: 0 },
+        { type: 'member', fromNode: 2, toNode: 6, delay: 0 },
+      ],
+    }
```
Comment (sz-piotr) [inline packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts:128]:
```suggestion
    const graph = [
      node(verifier, [edge('upgrade', 3)]),
      node(governance, [edge('upgrade', 3)]),
      node(zkSync, { threshold: 2 }, [
        edge('upgrade', 3),
        edge('configure', 2),
        edge('member', 5),
        edge('member', 6),
      ]),
      node(upgradeGatekeeper, [edge('act', 4)]),
      node(zkSyncMsig),
      node(actorA),
      node(actorB),
    ]
```

**Review by sz-piotr — APPROVED** [review]
